pkg/filter/or.go

```diff
@@ -0,0 +1,112 @@
+package filter
+
+import (
+	"fmt"
+	"sort"
+)
+
+// Or is a set of filters that should be evaluated as a logical
+// OR.
+type Or[T any] struct {
+	Filters []Filter[T]
+}
+
+func (o Or[T]) String() string {
+	s := "(or"
+	for _, f := range o.Filters {
+		s += fmt.Sprintf(" %s", f)
+	}
+
+	s += ")"
+	return s
+}
+
+// Flattened converts a filter into a minimal form, removing unnecessary
+// intermediate objects
+//
+// Flattened returns:
+// - AllCut if filter contains no filters
+// - AllPass if filter contains AllPass
+// - the inner filter if filter contains one filter
+// - an equivalent AllocationOr if filter contains more than one filter
+func (o Or[T]) Flattened() Filter[T] {
+	var flattenedFilters []Filter[T]
+	for _, innerFilter := range o.Filters {
+		flattenedInner := innerFilter.Flattened()
+
+		// Ignore AllCut in Or
+		if flattenedInner.equals(AllCut[T]{}) {
+			continue
+		}
+
+		// AllPass means the entire Or is AllPass
+		if flattenedInner.equals(AllPass[T]{}) {
+			return AllPass[T]{}
+		}
+
+		flattenedFilters = append(flattenedFilters, flattenedInner)
+	}
+
+	// Remove duplicates
+	flattenedFilters = removeDuplicates(flattenedFilters)
+
+	// Check for negations, if they exist And becomes AllPass
+	if checkForNegations(flattenedFilters) {
+		return AllPass[T]{}
+	}
+
+	// Empty Or is an AllPass
+	if len(flattenedFilters) == 0 {
+		return AllPass[T]{}
+	}
+
+	if len(flattenedFilters) == 1 {
+		return flattenedFilters[0]
+	}
+
+	// If more than half of the children are Not filters, Or should have DeMorgans rule applied to it
+	if percentNot(flattenedFilters) >= 0.5 {
```

Comment:
This appears to be different than the `And` filter condition to apply demorgan's rule. Shouldn't these be equivalent with respect to _when_ to apply the rule? See here: https://github.com/opencost/opencost/pull/1508#discussion_r1038569755

Reply:
I think you figured out already but this was about ensuring that conjunctions with even numbers of elements where half were negations, end up in the same format.

However I will be removing equality and flatten from filters to address this